CARDS-1736: Add ACL restriction for clinics

OrdiNeu opened the pull request:
[Jira link](https://phenotips.atlassian.net/browse/CARDS-1736)

This PR adds an ACL restriction that only applies if the user belongs to a group specified by the Visit Information form.

**To test**
1. Create a new clinic
![image](https://user-images.githubusercontent.com/4656440/172628106-6353498a-1abc-49a9-a29f-be9d37d16906.png)
2. Find the clinic in Composum under /Proms/ClinicMapping
![image](https://user-images.githubusercontent.com/4656440/172640921-359fcebb-a9b3-4679-983b-9307b60cb462.png)
3. Create a new Visit Information form, fill out the Clinic question with `/Proms/ClinicMapping/<id>` and save
![image](https://user-images.githubusercontent.com/4656440/172641166-f38af6b9-4247-4e5b-bd8b-42b10cac37f4.png)
4. Under Administration->User->Create a new test user
![image](https://user-images.githubusercontent.com/4656440/172641372-fcbd3f57-a155-41b8-b5fe-5a674f6ed14c.png)
5. Under Administration->Groups, add the new user to both TrustedUser and the clinic group generated in step 1
![image](https://user-images.githubusercontent.com/4656440/172641600-4eb9bd36-4c95-4966-bd51-72fe3aa265db.png)
6. Verify that you can see the Visit Information form from step 3 when signed in as your new test user
7. If you remove them from the clinic group in step 5, they will no longer be able to see the forms.
8. Visit Information forms without the cards:clinic set to `/Proms/ClinicMapping/<id>` should still be viewable.

OrdiNeu committed:
CARDS-1736: Add ACL restriction for clinics

marta- commented:
> QUESTION: As a test user, should I be able to change Clinic on the Visit information form?

Visit information forms will be pre-populated via EPR import and read-only for clinicians. Admins can technically do anything they want, but only a member of the technical team will have an admin account and they'll know not to mess with the data.

sashaandjic commented:
Problem found:
- created patient with visits with existing Cardio clinic and Mood clinic
- the I followed instruction from 1 - 6 ( my new clinic is Clinic A, of type Mood), created 2 visits with surveys
- when I logged as a test user, I was not able  to see only surveys that belonged to the Cardio clinic (Visit 1)
![screencapture-localhost-1736-content-html-Subjects-4075144e-1e9b-4ca4-b7af-054bff5f2200-2022-06-09-16_34_57](https://user-images.githubusercontent.com/23243244/172939024-eb3f5048-9ab3-4bf0-b18c-4d84e05c37f1.png)

sashaandjic commented:
> Visit information forms will be pre-populated via EPR import and read-only for clinicians. Admins can technically do anything they want, but only a member of the technical team will have an admin account and they'll know not to mess with the data.

Ok, that means my Test user should not be able to change it. Currently we allow it.

sashaandjic commented:
After following steps 1-8, I am unable to login as test user
![image](https://user-images.githubusercontent.com/23243244/172943882-d67820db-c929-4292-bd5b-8cefbad07b40.png)

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/permissions/ClinicRestrictionPattern.java:
```diff
@@ -0,0 +1,179 @@
+import io.uhndata.cards.dataentry.api.QuestionnaireUtils;
+
+/**
+ * A restriction that makes a permissions entry only be valid on a specific Clinic node.
```
```suggestion
 * A restriction that makes a permissions entry only be valid on a form for a specific Clinic.
```

marta- commented:
> > Visit information forms will be pre-populated via EPR import and read-only for clinicians. Admins can technically do anything they want, but only a member of the technical team will have an admin account and they'll know not to mess with the data.
> 
> Ok, that means my Test user should not be able to change it. Currently we allow it.

Indeed, it appears that nothing prevents TrustedUsers from editing data, which they really should not do. I suggest we deny `write` to all TrustedUsers (unless they created the form). Would that be enough or do we also need to specifically say the restrictions implemented here only grant read access? @sdumitriu

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/permissions/ClinicRestrictionPattern.java:
```diff
@@ -0,0 +1,179 @@
+    }
+
+    /**
+     * Return the visit information form for a given subject, or null if none exists.
```
```suggestion
     * Return the visit information form for a given subject, or null if it is inaccessible or none exists.
```

sdumitriu commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/permissions/ClinicRestrictionPattern.java:
```diff
@@ -0,0 +1,179 @@
+    public Resource getVisitInformationForSubject(String subjectID) throws RepositoryException
+    {
+        // Grab the uuid for the visit information questionnaire
+        Node visitInformationNode = this.session.getNodeByIdentifier(ClinicRestrictionPattern.VISIT_INFORMATION_PATH);
```
```suggestion
        Node visitInformationNode = this.session.getNode(ClinicRestrictionPattern.VISIT_INFORMATION_PATH);
```

sdumitriu committed:
CARDS-1736: Add ACL restriction for clinics

Minor comment and code improvements

sdumitriu committed:
CARDS-1736: Add ACL restriction for clinics
- Set the proper access rights to the forms subtree
- Made the restriction valueless, since it doesn't apply to a specific clinic
- Fixed StackOverflow error

OrdiNeu committed:
CARDS-1736: Look upwards for a form, if the current Tree is not one

OrdiNeu committed:
CARDS-1736: Major speedup by avoiding the Inner Join

sashaandjic commented:
In the restrictions dropdown for Principal `A123` (my test clinic) I do not see `cards:clinic`. When I selected `cards:clinicForms`, I was not able to see any forms from my test clinic A123, nor other clinic that also has surveys.
![image](https://user-images.githubusercontent.com/23243244/175766065-dabe0325-8216-4937-96b7-ad48c0af4f16.png)

sashaandjic commented:
Should `testuser` ever see anything other than Visit forms? I cannot see any surveys when I am logged as `testuser`.
![image](https://user-images.githubusercontent.com/23243244/175767150-c89db30d-221a-414f-8a51-4ac1b2e7f770.png)
![image](https://user-images.githubusercontent.com/23243244/175767192-ddaa16bc-9dd1-45e9-ae54-f1b86c771365.png)

OrdiNeu commented:
@sashaandjic 

> In the `Restrictions` dropdown for` Principal` `A123` (my test clinic) I do not see `cards:clinic`. !

I've removed this step from the testing instructions, as Sergiu's update now adds a `cards:clinicForms` ACL by default to `/Forms/`. 

> When I selected `cards:clinicForms` and logged as `testuser`, I was not able to see any forms from my test clinic A123, nor other clinic that also has surveys. 

This is probably related to the above, since the new default ACL works opposite of how I had it before.

To clarify the new steps:
1. Follow steps 1-8 in the PR description (I've removed the step that modifies the ACL)
2. Check that removing `testuser` from your new Clinic group in step 5 prevents them from seeing anything.

OrdiNeu committed:
CARDS-1736: Work with related subjects as well

marta- commented:
> Tested: Users can only see submitted survey forms and visit forms associated with the clinic they belong to ( by being a member of the clinic's group).
> 
> However, probably not related to this task, user can see visit object that belong to other clinics the user is not member of.

This should have been covered by this PR...

>Also, in the dashboard user can see all clinics, those that user is a member of and all others.
This is covered by #1084. We shoud probably rebase this branch on dev to include those changes, for a more realisting user experience while testing.

> 
> Loading patient chart is significantly slower now.

OrdiNeu committed:
CARDS-1736: Apply the clinic restriction pattern to subjects as well

OrdiNeu commented:
> However, probably not related to this task, user can see visit object that belong to other clinics the user is not member of. Also, in the dashboard user can see all clinics, those that user is a member of and all others.

Fixed in [0d19355](https://github.com/data-team-uhn/cards/pull/1055/commits/0d193558853bfa6ec87077ea5243458091971917)

sashaandjic commented:
Now my testuser cannot see any subjects, even those that are associated with the clinic the testuser is member of. I see in the dashboard 2 forms that I am supposed to see as testuser. Please note `undefines` in the place of subjectIDs.
![image](https://user-images.githubusercontent.com/23243244/176321156-cfa18bed-f555-418d-91ea-3164a680defe.png)
Opening a form ends up with WSOD.

Subject.jsx:132 Uncaught TypeError: Cannot read properties of undefined (reading 'label')
    at getHierarchy (Subject.jsx:132:139)
    at Form (Form.jsx:446:139)
    at renderWithHooks (react-dom.development.js:14985:18)
    at updateFunctionComponent (react-dom.development.js:17356:20)
    at beginWork (react-dom.development.js:19063:16)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945:14)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994:16)
    at invokeGuardedCallback (react-dom.development.js:4056:31)
    at beginWork$1 (react-dom.development.js:23959:7)
    at performUnitOfWork (react-dom.development.js:22774:12)
getHierarchy @ Subject.jsx:132
Form @ Form.jsx:446
renderWithHooks @ react-dom.development.js:14985
updateFunctionComponent @ react-dom.development.js:17356
beginWork @ react-dom.development.js:19063
callCallback @ react-dom.development.js:3945
invokeGuardedCallbackDev @ react-dom.development.js:3994
invokeGuardedCallback @ react-dom.development.js:4056
beginWork$1 @ react-dom.development.js:23959
performUnitOfWork @ react-dom.development.js:22774
workLoopSync @ react-dom.development.js:22702
renderRootSync @ react-dom.development.js:22665
performSyncWorkOnRoot @ react-dom.development.js:22288
eval @ react-dom.development.js:11327
unstable_runWithPriority @ scheduler.development.js:468
runWithPriority$1 @ react-dom.development.js:11276
flushSyncCallbackQueueImpl @ react-dom.development.js:11322
flushSyncCallbackQueue @ react-dom.development.js:11309
scheduleUpdateOnFiber @ react-dom.development.js:21888
dispatchAction @ react-dom.development.js:16139
handleResponse @ Form.jsx:315
Promise.then (async)
fetchData @ Form.jsx:300
eval @ Form.jsx:459
invokePassiveEffectCreate @ react-dom.development.js:23482
callCallback @ react-dom.development.js:3945
invokeGuardedCallbackDev @ react-dom.development.js:3994
invokeGuardedCallback @ react-dom.development.js:4056
flushPassiveEffectsImpl @ react-dom.development.js:23569
unstable_runWithPriority @ scheduler.development.js:468
runWithPriority$1 @ react-dom.development.js:11276
flushPassiveEffects @ react-dom.development.js:23442
performSyncWorkOnRoot @ react-dom.development.js:22264
eval @ react-dom.development.js:11327
unstable_runWithPriority @ scheduler.development.js:468
runWithPriority$1 @ react-dom.development.js:11276
flushSyncCallbackQueueImpl @ react-dom.development.js:11322
flushSyncCallbackQueue @ react-dom.development.js:11309
discreteUpdates$1 @ react-dom.development.js:22415
discreteUpdates @ react-dom.development.js:3756
dispatchDiscreteEvent @ react-dom.development.js:5889
react-dom.development.js:20085 The above error occurred in the <Form> component:

    at Form (webpack://cards-aggregated-frontend/./src/questionnaire/Form.jsx?:124:23)
    at C (webpack://cards-aggregated-frontend/./node_modules/react-router/esm/react-router.js?:725:37)
    at WithStyles (webpack://cards-aggregated-frontend/./node_modules/@mui/styles/withStyles/withStyles.js?:72:21)
    at Forms (webpack://cards-aggregated-frontend/./src/dataHomepage/Forms.jsx?:45:24)
    at WithStyles (webpack://cards-aggregated-frontend/./node_modules/@mui/styles/withStyles/withStyles.js?:72:21)
    at Page (webpack://cards-aggregated-frontend/./src/themePage/Page.jsx?:57:24)
    at Route (webpack://cards-aggregated-frontend/./node_modules/react-router/esm/react-router.js?:464:29)
    at Switch (webpack://cards-aggregated-frontend/./node_modules/react-router/esm/react-router.js?:670:29)
    at div
    at div
    at div
    at Suspense
    at div
    at Main (webpack://cards-aggregated-frontend/./src/themePage/index.jsx?:112:5)
    at WithStyles (webpack://cards-aggregated-frontend/./node_modules/@mui/styles/withStyles/withStyles.js?:72:21)
    at Route (webpack://cards-aggregated-frontend/./node_modules/react-router/esm/react-router.js?:464:29)
    at Switch (webpack://cards-aggregated-frontend/./node_modules/react-router/esm/react-router.js?:670:29)
    at Router (webpack://cards-aggregated-frontend/./node_modules/react-router/esm/react-router.js?:93:30)
    at InnerThemeProvider (webpack://cards-aggregated-frontend/./node_modules/@mui/system/esm/ThemeProvider/ThemeProvider.js?:22:67)
    at ThemeProvider (webpack://cards-aggregated-frontend/./node_modules/@mui/private-theming/ThemeProvider/ThemeProvider.js?:48:5)
    at ThemeProvider (webpack://cards-aggregated-frontend/./node_modules/@mui/system/esm/ThemeProvider/ThemeProvider.js?:42:5)
    at StyledEngineProvider (webpack://cards-aggregated-frontend/./node_modules/@mui/styled-engine/StyledEngineProvider/StyledEngineProvider.js?:24:5)

Consider adding an error boundary to your tree to customize error handling behavior.
Visit https://reactjs.org/link/error-boundaries to learn more about error boundaries.
logCapturedError @ react-dom.development.js:20085
update.callback @ react-dom.development.js:20113
callCallback @ react-dom.development.js:12318
commitUpdateQueue @ react-dom.development.js:12339
commitLifeCycles @ react-dom.development.js:20731
commitLayoutEffects @ react-dom.development.js:23421
callCallback @ react-dom.development.js:3945
invokeGuardedCallbackDev @ react-dom.development.js:3994
invokeGuardedCallback @ react-dom.development.js:4056
commitRootImpl @ react-dom.development.js:23146
unstable_runWithPriority @ scheduler.development.js:468
runWithPriority$1 @ react-dom.development.js:11276
commitRoot @ react-dom.development.js:22985
performSyncWorkOnRoot @ react-dom.development.js:22324
eval @ react-dom.development.js:11327
unstable_runWithPriority @ scheduler.development.js:468
runWithPriority$1 @ react-dom.development.js:11276
flushSyncCallbackQueueImpl @ react-dom.development.js:11322
flushSyncCallbackQueue @ react-dom.development.js:11309
scheduleUpdateOnFiber @ react-dom.development.js:21888
dispatchAction @ react-dom.development.js:16139
handleResponse @ Form.jsx:315
Promise.then (async)
fetchData @ Form.jsx:300
eval @ Form.jsx:459
invokePassiveEffectCreate @ react-dom.development.js:23482
callCallback @ react-dom.development.js:3945
invokeGuardedCallbackDev @ react-dom.development.js:3994
invokeGuardedCallback @ react-dom.development.js:4056
flushPassiveEffectsImpl @ react-dom.development.js:23569
unstable_runWithPriority @ scheduler.development.js:468
runWithPriority$1 @ react-dom.development.js:11276
flushPassiveEffects @ react-dom.development.js:23442
performSyncWorkOnRoot @ react-dom.development.js:22264
eval @ react-dom.development.js:11327
unstable_runWithPriority @ scheduler.development.js:468
runWithPriority$1 @ react-dom.development.js:11276
flushSyncCallbackQueueImpl @ react-dom.development.js:11322
flushSyncCallbackQueue @ react-dom.development.js:11309
discreteUpdates$1 @ react-dom.development.js:22415
discreteUpdates @ react-dom.development.js:3756
dispatchDiscreteEvent @ react-dom.development.js:5889
react-dom.development.js:67 Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
    at Form (webpack://cards-aggregated-frontend/./src/questionnaire/Form.jsx?:124:23)
    at C (webpack://cards-aggregated-frontend/./node_modules/react-router/esm/react-router.js?:725:37)
    at WithStyles (webpack://cards-aggregated-frontend/./node_modules/@mui/styles/withStyles/withStyles.js?:72:21)
    at Forms (webpack://cards-aggregated-frontend/./src/dataHomepage/Forms.jsx?:45:24)
    at WithStyles (webpack://cards-aggregated-frontend/./node_modules/@mui/styles/withStyles/withStyles.js?:72:21)
    at Page (webpack://cards-aggregated-frontend/./src/themePage/Page.jsx?:57:24)
    at Route (webpack://cards-aggregated-frontend/./node_modules/react-router/esm/react-router.js?:464:29)
    at Switch (webpack://cards-aggregated-frontend/./node_modules/react-router/esm/react-router.js?:670:29)
    at div
    at div
    at div
    at Suspense
    at div
    at Main (webpack://cards-aggregated-frontend/./src/themePage/index.jsx?:112:5)
    at WithStyles (webpack://cards-aggregated-frontend/./node_modules/@mui/styles/withStyles/withStyles.js?:72:21)
    at Route (webpack://cards-aggregated-frontend/./node_modules/react-router/esm/react-router.js?:464:29)
    at Switch (webpack://cards-aggregated-frontend/./node_modules/react-router/esm/react-router.js?:670:29)
    at Router (webpack://cards-aggregated-frontend/./node_modules/react-router/esm/react-router.js?:93:30)
    at InnerThemeProvider (webpack://cards-aggregated-frontend/./node_modules/@mui/system/esm/ThemeProvider/ThemeProvider.js?:22:67)
    at ThemeProvider (webpack://cards-aggregated-frontend/./node_modules/@mui/private-theming/ThemeProvider/ThemeProvider.js?:48:5)
    at ThemeProvider (webpack://cards-aggregated-frontend/./node_modules/@mui/system/esm/ThemeProvider/ThemeProvider.js?:42:5)
    at StyledEngineProvider (webpack://cards-aggregated-frontend/./node_modules/@mui/styled-engine/StyledEngineProvider/StyledEngineProvider.js?:24:5)
printWarning @ react-dom.development.js:67
error @ react-dom.development.js:43
warnAboutUpdateOnUnmountedFiberInDEV @ react-dom.development.js:23909
scheduleUpdateOnFiber @ react-dom.development.js:21835
dispatchAction @ react-dom.development.js:16139
handleFetchError @ Form.jsx:331
Promise.catch (async)
fetchData @ Form.jsx:300
eval @ Form.jsx:459
invokePassiveEffectCreate @ react-dom.development.js:23482
callCallback @ react-dom.development.js:3945
invokeGuardedCallbackDev @ react-dom.development.js:3994
invokeGuardedCallback @ react-dom.development.js:4056
flushPassiveEffectsImpl @ react-dom.development.js:23569
unstable_runWithPriority @ scheduler.development.js:468
runWithPriority$1 @ react-dom.development.js:11276
flushPassiveEffects @ react-dom.development.js:23442
performSyncWorkOnRoot @ react-dom.development.js:22264
eval @ react-dom.development.js:11327
unstable_runWithPriority @ scheduler.development.js:468
runWithPriority$1 @ react-dom.development.js:11276
flushSyncCallbackQueueImpl @ react-dom.development.js:11322
flushSyncCallbackQueue @ react-dom.development.js:11309
discreteUpdates$1 @ react-dom.development.js:22415
discreteUpdates @ react-dom.development.js:3756
dispatchDiscreteEvent @ react-dom.development.js:5889

sashaandjic commented:
I can reproduce the problem with `undefined` subjects on a brand new clone. It happens first time when I log as a test user. Restarting the cards instance seems to fix the problem.

Here are the findings with small data set. I have 2 clinics:
`ABC Clinic` - `test user` is part of `ABC group`
`CARDS-1736 Clinic`

I have 3 patients:
- `1234567` with 2 visits: one in each clinic
- `1111111` with 1 visit in `ABC clinic`
- `1111110` with 1 visit in` CARDS-1736 clinic`

Performance:
https://user-images.githubusercontent.com/23243244/176592618-39ec6b2b-0220-43b1-8612-1e8c8b06598e.mp4

Problem: Missing Patient Info form for patient 1234567 (has 1 visit with each clinic):
https://user-images.githubusercontent.com/23243244/176592729-c7ef43af-b9f4-4997-8c37-90853170be95.mp4
![image](https://user-images.githubusercontent.com/23243244/176592788-e038aedd-4f0d-41a9-81cd-7652975ee62d.png)

Trembling on the Patient chart page when scrolling down to the bottom of the page:
https://user-images.githubusercontent.com/23243244/176592944-4f9908c1-43d8-4c0b-ba10-ef9805b5b7ff.mp4

sashaandjic commented:
Performance with 200 patients - it takes 1min10sec to load questionnaires in the dashboard

https://user-images.githubusercontent.com/23243244/176596691-112caa2b-4a3d-42a6-968c-d52f34b08f9c.mp4

OrdiNeu committed:
CARDS-1736: Fix an ACL setting that was preventing users from seeing
patients of clinics they are part of

OrdiNeu commented:
@sashaandjic 
> I can reproduce the problem with `undefined` subjects on a brand new clone. It happens when I log as a test user. Restarting the cards instance seems to fix the problem.

Fixed in [ea12d20](https://github.com/data-team-uhn/cards/pull/1055/commits/ea12d2046f229f577972d6c397ade0aab3000c62) :heavy_check_mark:

sashaandjic commented:
All bugs that are reported have been fixed.

OrdiNeu commented:
Closing in favor of #1129
